rascaline/src/calculators/atomic_composition.rs

```diff
@@ -0,0 +1,269 @@
+
+        return Ok(samples);
+    }
+
+    fn supports_gradient(&self, parameter: &str) -> bool {
+        match parameter {
+            "positions" => true,
+            "cell" => true,
+            _ => false,
+        }
+    }
+
+    fn positions_gradient_samples(
+        &self,
+        keys: &Labels,
+        samples: &[Arc<Labels>],
+        systems: &mut [Box<dyn System>],
+    ) -> Result<Vec<Arc<Labels>>, Error> {
+        debug_assert_eq!(keys.count(), samples.len());
+        let mut gradient_samples = Vec::new();
+        for ([species_center_key], samples) in keys.iter_fixed_size().zip(samples) {
+            let mut builder = LabelsBuilder::new(vec!["sample", "structure", "atom"]);
+            if self.per_structure {
+                for (system_i, system) in systems.iter_mut().enumerate() {
+                    let species = system.species()?;
+
+                    for (center_i, &species_center_sys) in species.iter().enumerate() {
+                        if species_center_key.i32() == species_center_sys {
+                            builder.add(&[system_i, system_i, center_i]);
+                        }
+                    }
+                }
+            } else {
+                for (sample_i, [structure_i, center_i]) in samples.iter_fixed_size().enumerate() {
+                    builder.add(&[sample_i, structure_i.usize(), center_i.usize()]);
+                }
+            }
+            gradient_samples.push(Arc::new(builder.finish()));
+        }
+        return Ok(gradient_samples);
+    }
+
+    fn components(&self, keys: &Labels) -> Vec<Vec<Arc<Labels>>> {
+        return vec![Vec::new(); keys.count()];
+    }
+
+    fn properties_names(&self) -> Vec<&str> {
+        return vec!["count"];
+    }
+
+    fn properties(&self, keys: &Labels) -> Vec<Arc<Labels>> {
+        let mut properties = LabelsBuilder::new(self.properties_names());
+        properties.add(&[0]);
+        let properties = Arc::new(properties.finish());
+
+        return vec![properties; keys.count()];
+    }
+
+    fn compute(
+        &mut self,
+        systems: &mut [Box<dyn System>],
+        descriptor: &mut TensorMap,
+    ) -> Result<(), Error> {
+        assert_eq!(descriptor.keys().names(), ["species_center"]);
+
+        for (key, mut block) in descriptor.iter_mut() {
+            let values = block.values_mut();
+            let array = values.data.as_array_mut();
+
+            if self.per_structure {
+                let species = key[0].i32();
+                for (system_i, system) in systems.iter_mut().enumerate() {
+                    let all_species = system.species()?;
+                    let mut n_species = 0.;
+
+                    for &s in all_species {
+                        if s == species {
+                            n_species += 1.;
+                        }
+                    }
+                    array[[system_i, 0]] = n_species;
+                }
+            } else {
+                array.fill(1.0);
```

Comment:
this will fail is the user selects more samples than what this calculator would produce by default (i.e. trying to compute the composition of CH4 with 5 samples for center_species C (4 of which should be 0) and 5 samples for center_species H (1 of which should be 0)

Reply:
But, shouldn't this be taken of by the creation of the samples?

If no we do a loop like

```rust
for (sample_i, [system, center]) in values.samples.iter_fixed_size().enumerate() {
    for (property_i, property) in values.properties.iter().enumerate() {
        array[[sample_i, property_i]] = 1.0;
    }
}
```

Reply:
So this might need to be better documented, but the `samples` function returns the **default** set of samples. 

The actual ones for a given calculation might differ if the users passed something as `selected_samples` to `Calculator::compute`

Reply:
and yes, that's the typical loop one has to do, see all the other calculators